python/pyspark/worker.py

```diff
@@ -146,7 +146,52 @@ def verify_result_type(result):
     )
 
 
-def wrap_cogrouped_map_pandas_udf(f, return_type, argspec):
+def verify_pandas_result(result, return_type, assign_cols_by_name):
+    import pandas as pd
+
+    if not isinstance(result, pd.DataFrame):
+        raise TypeError(
+            "Return type of the user-defined function should be "
+            "pandas.DataFrame, but is {}".format(type(result))
+        )
+
+    # check the schema of the result only if it is not empty or has columns
+    if not result.empty or len(result.columns) != 0:
+        # if any column name of the result is a string
+        # the column names of the result have to match the return type
+        #   see create_array in pyspark.sql.pandas.serializers.ArrowStreamPandasSerializer
+        field_names = set([field.name for field in return_type.fields])
+        column_names = set(result.columns)
+        if (
+            assign_cols_by_name
+            and any(isinstance(name, str) for name in result.columns)
+            and column_names != field_names
+        ):
+            missing = sorted(list(field_names.difference(column_names)))
+            missing = "  Missing: " + (", ".join(missing)) if missing else ""
+
+            extra = sorted(list(column_names.difference(field_names)))
+            extra = "  Unexpected: " + (", ".join(extra)) if extra else ""
+
+            raise RuntimeError(
+                "Column names of the returned pandas.DataFrame do not match specified schema."
+                "{}{}".format(missing, extra)
+            )
+        # otherwise the number of columns of result have to match the return type
+        elif len(result.columns) != len(return_type):
+            raise RuntimeError(
+                "Number of columns of the returned pandas.DataFrame "
+                "doesn't match specified schema.  "
+                "Expected: {}  Actual: {}".format(
+                    len(return_type),
+                    len(result.columns)
```

Comment:
```suggestion
                "Number of columns of the returned pandas.DataFrame "
                "doesn't match specified schema.  "
                "Expected: {}.  Actual: {}".format(
                    len(return_type),
                    len(result.columns)
```

Reply:
Reverted this error message to before PR since that message is always short and easy to read without spaces or periods.